experimental/tools/cpp-simulation/cpp-simulation.cpp

```diff
@@ -0,0 +1,2237 @@
+  // ports
+  std::vector<ConsumerRW *> ins;
+  ProducerRW *outs;
+
+  // internal components
+  JoinSupport join;
+};
+
+class LazyForkModel : public OpExecutionModel<handshake::LazyForkOp> {
+public:
+  using OpExecutionModel<handshake::LazyForkOp>::OpExecutionModel;
+  LazyForkModel(handshake::LazyForkOp lazyForkOp,
+                mlir::DenseMap<Value, RW *> &subset)
+      : OpExecutionModel<handshake::LazyForkOp>(lazyForkOp), op(lazyForkOp),
+        size(op->getNumResults()),
+        ins(getState<ConsumerRW>(op.getOperand(), subset)), insData(ins) {
+
+    for (unsigned i = 0; i < size; ++i)
+      outs.push_back(getState<ProducerRW>(op->getResult(i), subset));
+
+    for (unsigned i = 0; i < size; ++i)
+      outsData.emplace_back(outs[i]);
+  }
+
+  void reset() override {
+    ins->ready = true;
+    for (unsigned i = 0; i < size; ++i) {
+      bool tempReady = true;
+      for (unsigned j = 0; j < size; ++j)
+        if (i != j)
+          tempReady = tempReady && outs[j]->ready;
+
+      ins->ready = ins->ready && outs[i]->ready;
+      outs[i]->valid = ins->valid && tempReady;
+    }
+    for (unsigned i = 0; i < size; ++i)
+      outsData[i] = insData;
+  }
+
+  void exec(bool isClkRisingEdge) override { reset(); }
+
+  void printStates() override {
+    printValue<ConsumerRW, const Data>("ins", ins, insData.data);
+    for (unsigned i = 0; i < size; ++i)
+      printValue<ProducerRW, Data>("outs", outs[i], outsData[i].data);
+  }
+
+private:
+  handshake::ForkOp op;
+
+  // parameters
+  unsigned size;
+
+  // ports
+  ConsumerRW *ins;
+  std::vector<ProducerRW *> outs;
+
+  ConsumerData insData;
+  std::vector<ProducerData> outsData;
+};
+
+class MergeModel : public OpExecutionModel<handshake::MergeOp> {
+public:
+  using OpExecutionModel<handshake::MergeOp>::OpExecutionModel;
+  MergeModel(handshake::MergeOp mergeOp, mlir::DenseMap<Value, RW *> &subset)
+      : OpExecutionModel<handshake::MergeOp>(mergeOp), op(mergeOp),
+        size(op->getNumOperands()),
+        outs(getState<ProducerRW>(op.getResult(), subset)), outsData(outs),
+        insTEHB(tehbValid, tehbReady) {
+    for (unsigned i = 0; i < size; ++i)
+      ins.push_back(getState<ConsumerRW>(op->getOperand(i), subset));
+
+    for (unsigned i = 0; i < size; ++i)
+      insData.emplace_back(ins[i]);
+    mergeTEHB.datawidth = outsData.dataWidth;
+  }
+
+  void reset() override {
+    if (outsData.hasValue())
+      resetDataFull(ins, outs, insData, outsData);
+    else
+      resetDataless(ins, outs);
+  }
+
+  void exec(bool isClkRisingEdge) override {
+    if (outsData.hasValue())
+      execDataFull(isClkRisingEdge, ins, outs, insData, outsData);
+    else
+      execDataless(isClkRisingEdge, ins, outs);
+  }
+
+  void printStates() override {
+    for (unsigned i = 0; i < size; ++i)
+      printValue<ConsumerRW, const Data>("ins", ins[i], insData[i].data);
+    printValue<ProducerRW, Data>("outs", outs, outsData.data);
+  }
+
+private:
+  handshake::MergeOp op;
```

Comment:
Noticing this here but this applies elsewhere, the `ExecutionModel` is already storing the operation automatically and `OpExecutionModel::getOperation` allows you to retrieve it with the correct type, so this is redundant information you should remove (it actually shadows the original field since it has the same name).